### PR TITLE
Namespace tidy

### PR DESCRIPTION
**Pull request by rhattersley**
This is an experiment in providing much more explicit control of the public API. The clarification is intended to help with simplifying any subsequent modification of the public API (deprecations, etc.), but also to remove irrelevant items from tab-completion (e.g. Jupyter notebook).

NB. The normal diff doesn't work very well, so I've kept the main code move & refactor as two bite-size commits.

### CODE REVIEW

File: biggus/_init.py [hunk 1]
```diff
@@ -66,7 +30,13 @@
...
 
 
-__version__ = '0.13.0'
+__all__ = []
+
+
+def export(defn):
+    globals()[defn.__name__] = defn
+    __all__.append(defn.__name__)
+    return defn
```
Comment (rhattersley) [inline biggus/_init.py]: Thanks to @dabeaz for this [neat little trick](http://www.dabeaz.com/modulepackage/ModulePackage.pdf).

**Comment by rhattersley** [comment]
So there we have it... the experiment works, in that the tests all pass and tab-completion works a treat. Does anyone have any thoughts on the style, implications, etc? Should we merge ... or not?

**Comment by pelson** [comment]
Done. The only implication which may be of concern is API incompatibility - for instance a user could have been modifying `biggus.MAX_CHUNK_SIZE` and that will now break. That is kind of the point of this PR though - expose the things that are supposed to be exposed, and make private the interfaces that weren't...

**Comment by djkirkham** [comment]
I know this is old now, but... I'm currently doing a piece of work where I need a masked biggus array. The solution (also suggested by @rhattersley) was to create my own subclass of `ArrayContainer`, but with this change I can't do that without importing `biggus._init`. I'll be submitting a PR with the subclass I created (which will hopefully be accepted!), but having ArrayContainer as part of the public api might be useful for other users.

**Comment by rhattersley** [comment]
:warning: Watch out if you import `biggus._init` - it's not part of the public API so could be subject to any change at any time. I'm not sure `ArrayContainer` has enough functionality and a clean enough API to warrant being exposed.

i.e. Submitting a PR that uses `ArrayContainer` is fine, but if you're writing your own code outside of biggus then I'd just subclass `Array`.

**Comment by djkirkham** [comment]
> but if you're writing your own code outside of biggus then I'd just subclass Array.

But then I have to implement all the methods myself. That's, like, 10 lines or something :cold_sweat:

**Comment by rhattersley** [comment]
:grinning:

**Comment by pelson** [comment]
@djkirkham - suggestion: put forward a PR making ArrayContainer public. You should get more detail regarding what is wrong with it, and we can either address the concerns, or factor parts of it into other *public* parts of the API.

**Pull request merged**